Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -0,0 +1,617 @@
+
+/-- The inductively constructed sequence `qₖ` in the proof of Weierstrass division. -/
+noncomputable def seq (H : g.IsWeierstrassDivisorAt I) (f : A⟦X⟧) : ℕ → A⟦X⟧
+  | 0 => 0
+  | k + 1 =>
+    H.seq f k + (PowerSeries.mk fun i ↦ coeff A (i + (g.map (Ideal.Quotient.mk I)).order.toNat)
+      (f - g * H.seq f k)) * H.isUnit_shift.unit⁻¹
+
+variable {f : A⟦X⟧}
+
+theorem coeff_seq_mem (k : ℕ) : ∀ i ≥ (g.map (Ideal.Quotient.mk I)).order.toNat,
+    coeff A i (f - g * H.seq f k) ∈ I ^ k := by
+  induction k with
+  | zero => simp
+  | succ k hq =>
+    rw [seq]
+    set q := H.seq f k
+    set s := f - g * q
+    set n := (g.map (Ideal.Quotient.mk I)).order.toNat
+    have hs := s.eq_X_pow_mul_shift_add_trunc n
+    set s₀ := s.trunc n
+    set s₁ := PowerSeries.mk fun i ↦ coeff A (i + n) s
+    set q' := q + s₁ * H.isUnit_shift.unit⁻¹
+    have key : f - g * q' = (s₀ : A⟦X⟧) - (g.trunc n : A⟦X⟧) * s₁ * H.isUnit_shift.unit⁻¹ := by
+      trans s + g * (q - q')
+      · simp_rw [s]; ring
+      simp_rw [q']
+      rw [sub_add_cancel_left, mul_neg, ← mul_assoc, mul_right_comm]
+      nth_rw 1 [g.eq_X_pow_mul_shift_add_trunc n]
+      rw [add_mul, mul_assoc, IsUnit.mul_val_inv, hs]
+      ring
+    intro i hi
+    rw [key, map_sub, Polynomial.coeff_coe, coeff_trunc, if_neg hi.not_lt, zero_sub, neg_mem_iff,
+      pow_succ']
+    refine coeff_mul_mem_ideal_of_coeff_left_mem_ideal' (fun i ↦ ?_) i
+    refine coeff_mul_mem_ideal_mul_ideal_of_coeff_mem_ideal'
+      (by simp [n, g.coeff_trunc_order_mem]) (fun i ↦ ?_) i
+    rw [coeff_mk]
+    exact hq (i + n) (by simp)
+
+theorem coeff_seq_succ_sub_seq_mem (k : ℕ) :
+    ∀ i, coeff A i (H.seq f (k + 1) - H.seq f k) ∈ I ^ k := by
+  simp_rw [seq, add_sub_cancel_left]
+  refine coeff_mul_mem_ideal_of_coeff_left_mem_ideal' fun i ↦ ?_
+  rw [coeff_mk]
+  exact H.coeff_seq_mem k (i + _) (by simp)
+
+@[simp]
+theorem seq_zero : H.seq f 0 = 0 := rfl
+
+theorem seq_one : H.seq f 1 = (PowerSeries.mk fun i ↦ coeff A
+    (i + (g.map (Ideal.Quotient.mk I)).order.toNat) f) * H.isUnit_shift.unit⁻¹ := by
+  simp_rw [seq, mul_zero, zero_add, sub_zero]
+
+variable (f)
+
+/-- The (bundled version of) coefficient of the limit `q` of the
+inductively constructed sequence `qₖ` in the proof of Weierstrass division. -/
+noncomputable def divCoeff [IsPrecomplete I A] (i : ℕ) :=
+  Classical.indefiniteDescription _ <| IsPrecomplete.prec' (I := I)
+    (fun k ↦ coeff A i (H.seq f k)) fun {m} {n} hn ↦ by
+      induction n, hn using Nat.le_induction with
+      | base => rfl
+      | succ n hn ih =>
+        refine ih.trans (SModEq.symm ?_)
+        simp_rw [SModEq.sub_mem, smul_eq_mul, Ideal.mul_top, ← map_sub]
+        exact Ideal.pow_le_pow_right hn (H.coeff_seq_succ_sub_seq_mem n i)
+
+/-- The limit `q` of the
+inductively constructed sequence `qₖ` in the proof of Weierstrass division. -/
+noncomputable def div [IsPrecomplete I A] : A⟦X⟧ := PowerSeries.mk fun i ↦ (H.divCoeff f i).1
```

Comment:
Same. Indeed all `def` that are only needed to prove the theorem should be `private`.

Reply:
No, later the `div` and `mod` will be used to construct the isomorphism `A[X] / (g) -> A[[X]] / (g)` (currently on my Iwasawa project). So at least I want to expose them as definitions and state some properties.

And note that some of them does not require full `IsAdicComplete`.